python/setup.py

```diff
@@ -114,87 +107,151 @@
                   "--singlegpu")
     sys.argv.remove('--multigpu')
 
-if "--singlegpu" in sys.argv:
-    cython_exc_list = glob.glob('cuml/*/*_mg.pyx')
-    cython_exc_list = cython_exc_list + glob.glob('cuml/*/*_mg.pxd')
-    cython_exc_list.append('cuml/nccl/nccl.pyx')
-    cython_exc_list.append('cuml/dask/common/comms_utils.pyx')
+if not libcuml_path:
+    libcuml_path = '../cpp/build/'
 
-    print('--singlegpu: excluding the following Cython components:')
-    pprint(cython_exc_list)
+##############################################################################
+# - Cython extensions build and parameters -----------------------------------
+#
+# We create custom build steps for both `build` and `build_ext` for several
+#   reasons:
+# 1) Custom `build_ext` is needed to set `cython_build_ext.cython_exclude` when
+#    `--singlegpu=True`
+# 2) Custom `build` is needed to exclude pacakges and directories when
+#    `--singlegpu=True`
+# 3) These cannot be combined because `build` is used by both `build_ext` and
+#    `install` commands and it would be difficult to set
+#    `cython_build_ext.cython_exclude` from `cuml_build` since the property
+#    exists on a different command.
+#
+# Using custom commands also allows combining commands at the command line. For
+# example, the following will all work as expected:
+# `python setup.py clean --all build --singlegpu build_ext --inplace`
+# `python setup.py clean --all build --singlegpu install --record=record.txt`
+# `python setup.py build_ext --debug --singlegpu`
 
-    python_exc_list = ["*.dask", "*.dask.*"]
-else:
-    libs.append('cumlprims')
-    libs.append('cumlcomms')
-    libs.append('nccl')
 
-    sys_include = os.path.dirname(sysconfig.get_path("include"))
-    include_dirs.append("%s/cumlprims" % sys_include)
+class cuml_build(_build):
 
-cmdclass = dict()
-cmdclass.update(versioneer.get_cmdclass())
-cmdclass["build_ext"] = build_ext
+    user_options = [
+        ("singlegpu", None, "Specifies whether to include multi-gpu or not")
+    ] + _build.user_options
 
-if not libcuml_path:
-    libcuml_path = '../cpp/build/'
+    boolean_options = ["singlegpu"] + _build.boolean_options
+
+    def initialize_options(self):
+
+        self.singlegpu = False
+
+        super().initialize_options()
+
+    def finalize_options(self):
+
+        # cumlcomms and nccl are still needed for multigpu algos not based
+        # on libcumlprims
+        libs = ['cuda', 'cuml++', 'rmm']
+
+        include_dirs = [
+            '../cpp/src',
+            '../cpp/include',
+            '../cpp/src_prims',
+            raft_include_dir,
+            '../cpp/comms/std/src',
+            '../cpp/comms/std/include',
+            cuda_include_dir,
+            numpy.get_include(),
+            os.path.dirname(sysconfig.get_path("include"))
+        ]
+
+        # Exclude multigpu components that use libcumlprims if
+        # --singlegpu is used
```

Comment:
Same as above, this was for the time when `--singlegpu` was not fully single GPU

Reply:
Done